core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAgentReflectConfigStep.java

```diff
@@ -0,0 +1,63 @@
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
+import io.quarkus.deployment.pkg.builditem.BuildSystemTargetBuildItem;
+import io.quarkus.deployment.pkg.steps.NativeOrNativeSourcesBuild;
+
+public class NativeImageAgentReflectConfigStep {
+    private static final Logger log = Logger.getLogger(NativeImageAgentReflectConfigStep.class);
+
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
+    void transformReflectConfig(BuildProducer<ReflectiveClassBuildItem> reflectClassProducer,
+            BuildProducer<ReflectiveMethodBuildItem> reflectMethodProducer,
+            BuildProducer<ReflectiveFieldBuildItem> reflectFieldProducer,
+            BuildSystemTargetBuildItem buildSystemTargetBuildItem) {
+        final Path agentReflectConfigPath = buildSystemTargetBuildItem.getOutputDirectory()
+                .resolve(Path.of("generated-native-config", "reflect-config.json"));
+        if (agentReflectConfigPath.toFile().exists()) {
+            try {
+                final String agentReflectConfig = Files.readString(agentReflectConfigPath);
+                final JsonReader jsonReader = JsonReader.of(agentReflectConfig);
+                jsonReader.<JsonReader.JsonArray> read().<JsonReader.JsonObject> stream()
+                        .forEach(reflectItem -> transformReflectItem(reflectItem, reflectClassProducer, reflectMethodProducer,
+                                reflectFieldProducer));
+            } catch (IOException e) {
+                log.debug("Unable to read file '" + agentReflectConfigPath.toAbsolutePath());
+            }
+        }
+    }
+
+    private void transformReflectItem(JsonReader.JsonObject reflectItem,
+            BuildProducer<ReflectiveClassBuildItem> reflectClassProducer,
+            BuildProducer<ReflectiveMethodBuildItem> reflectMethodProducer,
+            BuildProducer<ReflectiveFieldBuildItem> reflectFieldProducer) {
+        final JsonReader.JsonString name = reflectItem.get("name");
+        final JsonReader.JsonArray methods = reflectItem.get("methods");
+        if (methods != null) {
+            methods.<JsonReader.JsonObject> stream()
+                    .forEach(method -> transformReflectMethod(method, name.value(), reflectMethodProducer));
+        }
+        reflectClassProducer.produce(ReflectiveClassBuildItem.builder(name.value()).build());
+        // todo add fields
+    }
+
+    private void transformReflectMethod(JsonReader.JsonObject method, String className,
+            BuildProducer<ReflectiveMethodBuildItem> reflectMethodProducer) {
+        final String[] parameterTypes = method.<JsonReader.JsonArray> get("parameterTypes").<JsonReader.JsonString> stream()
+                .map(JsonReader.JsonString::value)
+                .toArray(String[]::new);
+        reflectMethodProducer.produce(
+                new ReflectiveMethodBuildItem(className, method.<JsonReader.JsonString> get("name").value(), parameterTypes));
```

Comment:
Why the `json -> BuildItem -> json` step? Why not just merge the json files?

Reply:
Hmmm, care to share more details? I don't see how it would be vastly different or better than this.